File: Activities/InternetDatabaseSearchActivity.cs
using System;
using System.IO;
using System.Net;
using System.Threading;
using System.Globalization;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.OS;
using Android.Support.V4.Content;
using Android.Views;
using Android.Widget;
using Android.Content.PM;

using Newtonsoft.Json;

namespace VorratsUebersicht
{
    using static Tools;

    [Activity(Label = "https://openfoodfacts.org")]
    public class InternetDatabaseSearchActivity : Activity
    {
        FoodInformation foodInfo = null;
        QuantityAndUnit foodSize = null;
        decimal? kcalPer100 = null;

        public static Bitmap picture = null;
        public string formatedResponseFromServer = string.Empty;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            // ActionBar Hintergrund Farbe setzen
            var backgroundPaint = ContextCompat.GetDrawable(this, Resource.Color.Application_ActionBar_Background);
            backgroundPaint.SetBounds(0, 0, 10, 10);
            ActionBar.SetBackgroundDrawable(backgroundPaint);
            ActionBar.SetDisplayHomeAsUpEnabled(true);

            this.SetContentView(Resource.Layout.InternetDatabaseSearch);

            string eanCode = Intent.GetStringExtra ("EANCode") ?? string.Empty;

            var textView = FindViewById<TextView>(Resource.Id.InternetDatabaseResult_ProgressText);
            textView.Text = this.Resources.GetString(Resource.String.InternetDatabaseSearch_Searching);

            var textInfo = FindViewById<TextView>(Resource.Id.InternetDatabaseResult_Description);
            textInfo.Text = this.Resources.GetString(Resource.String.InternetDatabaseSearch_Info);

            new System.Threading.Thread(new ThreadStart(delegate             
            {
                this.SearchAndShowArticle(eanCode);
            })).Start();

        }

        public override bool OnCreateOptionsMenu(IMenu menu)
        {
            MenuInflater.Inflate(Resource.Menu.InternetDatabaseSearch_menu, menu);
            return base.OnCreateOptionsMenu(menu);
        }

        public override bool OnOptionsItemSelected(IMenuItem item)
        {
            switch(item.ItemId)
            {
                case Android.Resource.Id.Home:
                    this.OnBackPressed();
                    return true;

                case Resource.Id.InternetDatabaseResult_Save:

                    if (this.foodInfo?.status == 1)
                    {
                        Intent intent = new Intent();
                        intent.PutExtra("Name",       this.foodInfo.product.GetProduktName());
                        intent.PutExtra("Hersteller", this.foodInfo.product.brands);
                        if (this.foodSize != null)
                        {
                            intent.PutExtra("Quantity",   (long)this.foodSize.Quantity);
                            intent.PutExtra("Unit",       this.foodSize.Unit);
                        }
                        if (this.kcalPer100 != null)
                        {
                            intent.PutExtra("KCalPer100",   (long)this.kcalPer100.Value);
                            
                        }

                        this.SetResult(Result.Ok, intent);
                    }
                    else
                    {
                        this.SetResult(Result.Canceled);
                        InternetDatabaseSearchActivity.picture = null;
                    }

                    this.OnBackPressed();
                    return true;

            }

            return false;
        }


        private void SearchAndShowArticle(string eanCode)
        {
            /*
            foodInfo.product = new FoodInformation.Product();
            foodInfo.product.image_url       = "https://static.openfoodfacts.org/images/products/20005016/front_fr.4.400.jpg";
            */

            //eanCode = "22120649";   // Kartoffeleintopf Mit Würstchen Und Rauchspeck, mit Mengenangabe, und KCal/100g
            //eanCode = "88888888888";  // Not Found
            //eanCode = "4000462810052";  // Goldpuder Weizenmehl Type 405 1 KG
            //eanCode = "5410673854001";  // Reis, mit Einheit und Nährstoff

            string title = string.Empty;
            string info = string.Empty;

            InternetDatabaseSearchActivity.picture = null;
            this.formatedResponseFromServer = string.Empty;
            this.foodInfo = null;
            this.foodSize = null;
            this.kcalPer100 = null;

            var eanCodeView  = FindViewById<TextView>(Resource.Id.InternetDatabaseResult_EanCode);
            var progressText = FindViewById<TextView>(Resource.Id.InternetDatabaseResult_ProgressText);
            var progressBar  = FindViewById<ProgressBar>(Resource.Id.InternetDatabaseResult_Progress);
            
            try
            {
                string[] eanCodeList = eanCode.Split(",");
                foreach(string ean in eanCodeList)
                {
                    RunOnUiThread(() => eanCodeView.Text = string.Format(this.Resources.GetString(Resource.String.InternetDatabaseSearch_EANCode), ean));
                    
                    this.foodInfo = this.GetFoodInformation(ean);

                    if (foodInfo.status == 1)       // Artikelangaben gefunden?
                        break;
                }

            }
            catch(Exception ex)
            {
                TRACE(ex);

                title = ex.Message;

                if (ex.InnerException != null)
                {
                    TRACE(ex.InnerException);
                }

            }

            RunOnUiThread(() =>
            {
                eanCodeView.Text = string.Empty;

                if (this.foodInfo != null)
                {
                    title = string.Format(this.Resources.GetString(Resource.String.InternetDatabaseSearch_EANCode), this.foodInfo.code);

                    if (this.foodInfo.status == 1)
                    {
                        info += string.Format("{0}\n{1}\n\n", this.Resources.GetString(Resource.String.ArticleDetails_ArticleName),  this.foodInfo.product.GetProduktName());
                        info += string.Format("{0}\n{1}\n\n", this.Resources.GetString(Resource.String.ArticleDetails_Manufacturer), this.foodInfo.product.brands);

                        this.foodSize = QuantityAndUnit.Parse(this.foodInfo.product.quantity);
                        if (this.foodSize != null)
                        {
                            string content = this.Resources.GetString(Resource.String.InternetDatabaseSearch_Size);
                            info += string.Format(CultureInfo.CurrentUICulture, content,  this.foodSize.Quantity, this.foodSize.Unit);
                        }
                        else
                        {
                            string unknownSize = this.Resources.GetString(Resource.String.InternetDatabaseSearch_UnknownSize);
                            info += string.Format(unknownSize, this.foodInfo.product.quantity);
                        }

                        if (foodInfo.product.nutriments != null)
                        {
                            /*
                            info += string.Format("Nährwert: {0} {1} pro 100g\n", 
                                this.foodInfo.product.nutriments.energy_100g,
                                this.foodInfo.product.nutriments.energy_unit);
                            */

                            string caloriePer100gInfo = this.Resources.GetString(Resource.String.InternetDatabaseSearch_CaloriesPerUnit);

                            if (string.Compare(this.foodInfo.product.nutriments.energy_unit, "kcal", true) == 0)
                            {
                                this.kcalPer100 = this.foodInfo.product.nutriments.energy_value;
                                info += "\n\n";
                                info += string.Format(caloriePer100gInfo, this.kcalPer100);
                            }

                            if (string.Compare(this.foodInfo.product.nutriments.energy_unit, "kJ", true) == 0)
                            {
                                // kcal = kJ / 4,184 dividieren
                                this.kcalPer100 = this.foodInfo.product.nutriments.energy_100g / 4.184m;
                                this.kcalPer100 = Math.Round(this.kcalPer100.Value, 0);

                                info += "\n\n";
                                info += string.Format(caloriePer100gInfo, this.kcalPer100);
                            }
                        }

                        var languageCode = CultureInfo.CurrentCulture.TwoLetterISOLanguageName;

                        eanCodeView.Text = "Link: https://" + languageCode + ".openfoodfacts.org/product/" + this.foodInfo.code;
                    }
                    else
                    {
                        info += string.Format("Status: {0} - {1}\n\n", this.foodInfo.status, this.foodInfo.status_verbose);
                        info += string.Format("Response: {0}",   this.formatedResponseFromServer);
                    }
                }

                progressText.Text = title;
                progressBar.Visibility = ViewStates.Gone;

                var textView = FindViewById<TextView>(Resource.Id.InternetDatabaseResult_Text);
                if (textView == null)
                    return;

                textView.Text = info;

                if (this.foodInfo?.product != null)
                {
                    InternetDatabaseSearchActivity.picture = this.GetUrlPicture(this.foodInfo.product.image_url);

                    var imageView = FindViewById<ImageView>(Resource.Id.InternetDatabaseResult_Image);
                    imageView.SetImageBitmap(InternetDatabaseSearchActivity.picture);
                }
            });
        }

        private FoodInformation GetFoodInformation(string eanCode)
        {
            string parameter;
            
            parameter = "?fields=product_name,product_name_de,brands,quantity,nutriments,image_url";

            string server = "https://world.openfoodfacts.org/api/v0/product/";
            string request = string.Format("{0}{1}.json", server, eanCode);
            if (!string.IsNullOrEmpty(parameter))
            {
                request += parameter;
            }

            TRACE("WebRequest: {0}", request);
            Context context = this.ApplicationContext;
            PackageInfo info = context.PackageManager.GetPackageInfo(context.PackageName, 0);

            string agentInfo = "Vorratsübersicht - Android - Version " + info.VersionName + " - https://sites.google.com/site/vorratsuebersicht";

            WebRequest webRequest = WebRequest.Create(request);
            webRequest.Credentials = CredentialCache.DefaultCredentials;
            webRequest.Headers.Add("UserAgent", agentInfo);
            webRequest.Timeout = 10000;

            if (ServicePointManager.ServerCertificateValidationCallback == null)
            {
                ServicePointManager.ServerCertificateValidationCallback += 
                    (sender, certificate, chain, sslPolicy) => { return true; };
            }

            WebResponse response = webRequest.GetResponse();

            string webResponse = string.Empty;

            using (Stream dataStream = response.GetResponseStream())
            {
                StreamReader reader = new StreamReader(dataStream);
                webResponse = reader.ReadToEnd();
                TRACE("Response from OpenFoodFacts.org (max 1024 Zeichen):\n{0}", webResponse);
            }
            
            response.Close();

            var foodInfo = JsonConvert.DeserializeObject<FoodInformation>(webResponse);

            // Formatierte Ausgabe
            var allInfo = JsonConvert.DeserializeObject(webResponse);
            this.formatedResponseFromServer = JsonConvert.SerializeObject(allInfo, Formatting.Indented);
            TRACE("JSon Antwort vom Server OpenFoodFacts.org:\n\n{0}", this.formatedResponseFromServer);

            return foodInfo;
        }

        private Bitmap GetUrlPicture(string imageUrl)
        {
            Bitmap bitmap = null;

            if (string.IsNullOrEmpty(imageUrl))
                return bitmap;

            WebRequest webRequest = WebRequest.Create(imageUrl);
            webRequest.Credentials = CredentialCache.DefaultCredentials;

            WebResponse response = webRequest.GetResponse();

            byte[] bytes = null;

            using (Stream dataStream = response.GetResponseStream())
            using( MemoryStream ms = new MemoryStream())
            {
                int count = 0;
                do
                {
                    byte[] buf = new byte[1024];
                    count = dataStream.Read(buf, 0, 1024);
                    ms.Write(buf, 0, count);
                } while(dataStream.CanRead && count > 0);

                bytes = ms.ToArray();
            }
            bitmap = BitmapFactory.DecodeByteArray(bytes, 0, bytes.Length);
            response.Close();

            return bitmap;
        }
    }

    public class FoodInformation
    {
        public string code;
        public int status;
        public string status_verbose;
        public Product product;

        public class Product
        {
            public string product_name;
            public string product_name_de;
            public string brands;
            public string code;
            public string image_url;

            public string quantity;
            public int product_quantity;

            // Nährstoffe
            public Nutriments nutriments;

            internal string GetProduktName()
            {
                // Vorzugsweise die Deutsche Sprache nehmen. Beispiel: https://world.openfoodfacts.org/api/v0/product/7610200068989.json
                if (!string.IsNullOrEmpty(this.product_name_de))
                    return this.product_name_de;

                return this.product_name;
            }
        }

        public class Nutriments
        {
            public string energy_unit;
            public decimal energy_value;  // kcal pro 100 g oder 100 ml
            public decimal energy_100g;   // {energy_unit} pro 100 g oder 100 ml  
        }
    }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/InternetDatabaseSearchActivity.kt
package de.stryi.vorratsuebersicht

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.menu.MenuBuilder
import androidx.lifecycle.lifecycleScope
import com.google.gson.Gson
import de.stryi.vorratsuebersicht.databinding.InternetDatabaseSearchBinding
import de.stryi.vorratsuebersicht.tools.QuantityAndUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.util.Locale


class InternetDatabaseSearchActivity :AppCompatActivity()
{
    private lateinit var binding: InternetDatabaseSearchBinding

    private var formatedResponseFromServer : String = ""

    var foodInfo: FoodInformation? = null
    var foodSize: QuantityAndUnit? = null
    var kcalPer100 : Double? = null

    companion object {
        var picture: Bitmap? = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = InternetDatabaseSearchBinding.inflate(layoutInflater)

        setContentView(binding.root)

        this.setSupportActionBar(binding.InternetDatabaseResultAppBar)

        binding.InternetDatabaseResultAppBar.setNavigationOnClickListener { finish() }

        binding.InternetDatabaseResultProgressText.setText(R.string.InternetDatabaseSearch_Searching)
        binding.InternetDatabaseResultDescription.setText(R.string.InternetDatabaseSearch_Info)

        val eanCode = intent.getStringExtra("EANCode")

        if (eanCode == null)
        {
            binding.InternetDatabaseResultProgress.visibility = View.INVISIBLE
            binding.InternetDatabaseResultText.text = "Kein EAN Code angegeben."
            return
        }

        binding.InternetDatabaseResultEanCode.text = this.resources.getString(R.string.InternetDatabaseSearch_EANCode).format(eanCode)

        // TODO: eanCode könnte auch mehrere enthalten (Komma seppariert)
        this.searchAndShowArticle(eanCode)
    }

    @SuppressLint("RestrictedApi")
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (menu is MenuBuilder) {
            menu.setOptionalIconsVisible(true)
        }
        menuInflater.inflate(R.menu.internet_database_search_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.InternetDatabaseResult_Save -> {

                if (foodInfo?.status == 1)
                {
                    val intent = Intent()
                    intent.putExtra("Name", foodInfo?.product?.getProduktName())
                    intent.putExtra("Hersteller", foodInfo?.product?.brands)
                    if (foodSize != null)
                    {
                        intent.putExtra("Quantity", foodSize?.quantity)
                        intent.putExtra("Unit",     foodSize?.unit)
                    }
                    if (kcalPer100 != null) {
                        intent.putExtra("KCalPer100", kcalPer100!!)
                    }

                    this.setResult(RESULT_OK, intent)
                }
                else
                {
                    this.setResult(RESULT_CANCELED)
                }

                this.finish()
                return true
            }
        }

        return false
    }

    private fun searchAndShowArticle(eanCode : String) {

        lifecycleScope.launch {
            try {
                // Führe die Netzwerkanfrage im IO-Dispatcher (Hintergrund) aus
                foodInfo = withContext(Dispatchers.IO) {
                    getFoodInformation(eanCode)
                }

                foodSize   = QuantityAndUnit.parse(foodInfo?.product?.quantity)
                kcalPer100 = getKcalPer100(foodInfo)

                picture = withContext(Dispatchers.IO) {
                    getUrlPicture(foodInfo?.product?.image_url)
                }

                if (picture != null)
                {
                    binding.InternetDatabaseResultImage.setImageBitmap(picture)
                    binding.InternetDatabaseResultImage.visibility = View.VISIBLE
                }

                binding.InternetDatabaseResultProgress.visibility = View.INVISIBLE

                // Wechsle zurück zum Main-Thread, um die UI zu aktualisieren
                if (foodInfo == null || foodInfo?.status == 0)   // OpenFoodFacts gibt status=0 bei Nicht-Fund
                {
                    val status   = "Status: %s - %s\n\nResponse: %s".format(
                        foodInfo?.status,
                        foodInfo?.status_verbose,
                        formatedResponseFromServer)

                    binding.InternetDatabaseResultText.text = status


                }
                else
                {
                    var info = resources.getString(R.string.ArticleDetails_ArticleName)
                    info += "\n%s\n\n".format(foodInfo?.product?.getProduktName())

                    info += resources.getString(R.string.ArticleDetails_Manufacturer)
                    info += "\n%s\n\n".format(foodInfo?.product?.brands)

                    if (foodSize != null)
                    {
                        info += resources.getString(R.string.InternetDatabaseSearch_Size)
                            .format(foodSize?.quantity, foodSize?.unit)
                    }
                    else
                    {
                        info += resources.getString(R.string.InternetDatabaseSearch_UnknownSize)
                            .format(foodInfo?.product?.quantity)
                    }

                    if (kcalPer100 != null)
                    {
                        info += "\n\n"
                        info += resources.getString(R.string.InternetDatabaseSearch_CaloriesPerUnit)
                            .format(kcalPer100)
                    }

                    binding.InternetDatabaseResultText.text = info
                    val languageCode = Locale.getDefault().language
                    binding.InternetDatabaseResultEanCode.text =
                        "Link: https://$languageCode.openfoodfacts.org/product/" + foodInfo?.code
                }

            } catch (e: Exception) {
                // Fehlerbehandlung bleibt auf dem Main-Thread
                binding.InternetDatabaseResultProgress.visibility = View.INVISIBLE
                binding.InternetDatabaseResultText.text = "Fehler bei der Suche: ${e.message}"
            }
        }
    }

    private fun getFoodInformation(eanCode: String): FoodInformation? {

        val fields = "product_name,product_name_de,brands,quantity,nutriments,image_url"

        val url = "https://world.openfoodfacts.org/api/v0/product/$eanCode.json?fields=$fields"

        val info = this.packageManager.getPackageInfo(this.packageName, 0)
        val userAgent = ("Vorratsuebersicht - Android - Version " + info.versionName + " - https://sites.google.com/site/vorratsuebersicht")

        val client = OkHttpClient().newBuilder()
            .connectTimeout(10, java.util.concurrent.TimeUnit.SECONDS)
            .build()

        val request = Request.Builder()
            .header("User-Agent", userAgent)
            .url(url)
            .build()

        val gson = Gson()

        val call = client.newCall(request)
        val response = call.execute()
        try {
            if (!response.isSuccessful) {
                throw IOException("Unexpected code $response")
            }

            val bodyString = response.body?.string()
            if (bodyString == null) {
                return null
            }
            this.formatedResponseFromServer = bodyString

            return gson.fromJson(bodyString, FoodInformation::class.java)
        } finally {
            response.close()  // wichtig: Ressourcen freigeben
        }
    }

    data class FoodInformation(
        val code: String?,
        val status: Int?,
        val status_verbose: String?,
        val product: Product?
    ) {
        data class Product(
            val product_name: String?,
            val product_name_de: String?,
            val brands: String?,
            val code: String?,
            val image_url: String?,
            val quantity: String?,
            val product_quantity: Int?,
            val nutriments: Nutriments?
        ) {
            fun getProduktName(): String? {
                // Vorzugsweise die Deutsche Sprache nehmen
                return if (!product_name_de.isNullOrEmpty()) {
                    product_name_de
                } else {
                    product_name
                }
            }
        }

        data class Nutriments(
            val energy_unit: String?,
            val energy_value: Double?,   // kcal pro 100 g oder 100 ml
            val energy_100g: Double?     // {energy_unit} pro 100 g oder 100 ml
        )
    }

    fun getKcalPer100(foodInfo: FoodInformation?): Double?
    {
        var value : Double? = null

        if (foodInfo?.product?.nutriments?.energy_unit.equals("kcal", ignoreCase = true)) {
            value = foodInfo?.product?.nutriments?.energy_value
        }

        if (foodInfo?.product?.nutriments?.energy_unit.equals("kJ", ignoreCase = true)) {
            // kcal = kJ / 4,184 dividieren
            value = foodInfo?.product?.nutriments?.energy_100g?.div(4.184)
            value = value?.let { kotlin.math.round(it) }
        }
        return value
    }

    fun getUrlPicture(imageUrl : String?) : Bitmap?
    {
        if (imageUrl == null)
            return null

        val client = OkHttpClient
            .Builder()
            .build()
        val request = Request.Builder()
            .url(imageUrl)
            .build()
        val call = client.newCall(request)
        val response = call.execute()
        try {
            if (!response.isSuccessful) {
                throw IOException("Unexpected code $response")
            }

            val body = response.body?.bytes()
            if (body == null) {
                return null
            }
            return BitmapFactory.decodeByteArray(body, 0, body.size)
        } finally {
            response.close()  // wichtig: Ressourcen freigeben
        }
    }
}